include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp

```diff
@@ -64,23 +82,35 @@ namespace nil {
                             std::swap(a[k], a[rk]);
                     }
 
-                    std::size_t m = 1;    // invariant: m = 2^{s-1}
-                    for (std::size_t s = 1; s <= logn; ++s) {
+                    for (std::size_t s = 1, m = 1; s <= logn; ++s, m *= 2) { // invariant: m = 2^{s-1}
                         // w_m is 2^s-th root of unity now
-                        const field_value_type w_m = omega.pow(n / (2 * m));
-
+                        const std::size_t inc = n / (2 * m);
                         asm volatile("/* pre-inner */");
                         for (std::size_t k = 0; k < n; k += 2 * m) {
-                            field_value_type w = field_value_type::one();
-                            for (std::size_t j = 0; j < m; ++j) {
-                                const value_type t = a[k + j + m] * w;
+                            for (std::size_t j = 0, idx = 0; j < m; ++j, idx += inc) {
+                                const value_type t = a[k + j + m] * omega_cache[idx];
```

Comment:
Let's try one more thing, check if it helps with performance. We are creating a variable "t" here, which takes time with out multiprecision.

Let's create it somewhere outside this loops, something like:
                     value_type t;
                     for (std::size_t s = 1, m = 1; s <= logn; ++s, m *= 2) { // invariant: m = 2^{s-1}
                        // w_m is 2^s-th root of unity now
                        const std::size_t inc = n / (2 * m);
                        asm volatile("/* pre-inner */");
                        for (std::size_t k = 0; k < n; k += 2 * m) {
                            for (std::size_t j = 0, idx = 0; j < m; ++j, idx += inc) {
                                t = std::move(a[k + j + m]);
                                t *= omega_cache[idx];
                                a[k + j + m] = a[k + j];
                                a[k + j + m] -= t;
                                a[k + j] += t;
                            }
                        }
                        asm volatile("/* post-inner */");

This is approximate code, if it does not work contact me.

Reply:
Worked nicely, and was 3-5% faster compared to the old optimized version.